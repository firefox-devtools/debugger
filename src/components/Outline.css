.outline-list {
  list-style-type: "-";
}

.outline-list__element {
  color: blue;
  padding-left: 0.5rem;
}
