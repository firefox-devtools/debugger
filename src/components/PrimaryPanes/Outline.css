.outline {
  overflow-y: auto;
}

.outline .outline-pane-info {
  width: 100%;
  font-style: italic;
  text-align: center;
  padding: 0.5em;
  user-select: none;
  font-size: 12px;
}

.outline-list {
  list-style-type: none;
  padding-left: 0px;
  width: 100%;
}

.outline-list__element {
  color: blue;
  padding-bottom: 0.2rem;
  padding-left: 1rem;
  padding-right: 0.5rem;
  padding-top: 0.2rem;
  cursor: default;
}

.outline-list__element:hover {
  background: var(--theme-toolbar-background-hover);
}
