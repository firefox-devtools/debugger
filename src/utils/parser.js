// @flow

const babylon = require("babylon");
const traverse = require("babel-traverse").default;
const t = require("babel-types");
const { isDevelopment } = require("devtools-config");
const toPairs = require("lodash/toPairs");
const get = require("lodash/get");
const isEmpty = require("lodash/isEmpty");

import type { SourceText, Location } from "../types";

type ASTLocation = {
  start: {
    line: number,
    column: number
  },
  end: {
    line: number,
    column: number
  }
};

export type SymbolDeclaration = {
  name: string,
  location: ASTLocation
};

export type FormattedSymbolDeclaration = {
  id: string,
  title: string,
  subtitle: string,
  value: string,
  location: ASTLocation
};

export type SymbolDeclarations = {
  functions: Array<FormattedSymbolDeclaration>,
  variables: Array<FormattedSymbolDeclaration>,
  classes: Array<FormattedSymbolDeclaration>,
};

const ASTs = new Map();

const symbolDeclarations = new Map();

function _parse(code) {
  return babylon.parse(code, {
    sourceType: "module",

    plugins: [
      "jsx",
      "flow"
    ]
  });
}

function parse(text: string) {
  let ast;
  if (!text) {
    return;
  }

  try {
    ast = _parse(text);
  } catch (error) {
    if (isDevelopment()) {
      console.warn("parse failed", text);
    }

    ast = {};
  }

  return ast;
}

function getAst(sourceText: SourceText) {
  if (ASTs.has(sourceText.id)) {
    return ASTs.get(sourceText.id);
  }

  const ast = parse(sourceText.text);
  ASTs.set(sourceText.id, ast);
  return ast;
}

function getFunctionName(path) {
  if (path.node.id) {
    return path.node.id.name;
  }

  const parent = path.parent;
  if (parent.type == "ObjectProperty") {
    return parent.key.name;
  }

  if (parent.type == "ObjectExpression" || path.node.type == "ClassMethod") {
    return path.node.key.name;
  }

  if (parent.type == "VariableDeclarator") {
    return parent.id.name;
  }

  return "anonymous";
}

function isFunction(path) {
  return t.isFunction(path) || t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) || t.isClassMethod(path);
}

function formatSymbol(symbol: SymbolDeclaration): FormattedSymbolDeclaration {
  return {
    id: `${symbol.name}:${symbol.location.start.line}`,
    title: symbol.name,
    subtitle: `:${symbol.location.start.line}`,
    value: symbol.name,
    location: symbol.location
  };
}

function getVariableNames(path) {
  if (t.isObjectProperty(path) && !isFunction(path.node.value)) {
    return [formatSymbol({
      name: path.node.key.name,
      location: path.node.loc
    })];
  }

  if (!path.node.declarations) {
    return path.node.params
    .map(dec => formatSymbol({
      name: dec.name,
      location: dec.loc
    }));
  }

  return path.node.declarations
    .map(dec => formatSymbol({
      name: dec.id.name,
      location: dec.loc
    }));
}

function isVariable(path) {
  return t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value));
}

function getMemberExpression(root) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }
    return [node.name].concat(expr);
  }

  return _getMemberExpression(root, []);
}

function getSymbols(source: SourceText): SymbolDeclarations {
  if (symbolDeclarations.has(source.id)) {
    const symbols = symbolDeclarations.get(source.id);
    if (symbols) {
      return symbols;
    }
  }

  const ast = getAst(source);
  const symbols = { functions: [], variables: [], classes: [] };

  traverse(ast, {
    enter(path) {
      if (isVariable(path)) {
        symbols.variables.push(...getVariableNames(path));
      }

      if (isFunction(path)) {
        symbols.functions.push(formatSymbol({
          name: getFunctionName(path),
          location: path.node.loc
        }));
      }

      if (t.isClassDeclaration(path)) {
        symbols.classes.push(formatSymbol({
          name: path.node.id.name,
          location: path.node.loc
        }));
      }
    }
  });

  symbolDeclarations.set(source.id, symbols);
  return symbols;
}

function getExpression(source: SourceText, token: string, location: Location) {
  let expression = null;
  const ast = getAst(source);

  if (isEmpty(ast)) {
    return;
  }

  traverse(ast, {
    enter(path) {
      const node = path.node;
      if (t.isMemberExpression(node) && node.property.name === token
        && nodeContainsLocation({ node, location })) {
        const expr = getMemberExpression(node);
        expression = {
          value: expr.join("."),
          location: node.loc
        };
      }
    }
  });

  return expression;
}

function nodeContainsLocation({ node, location }) {
  const { start, end } = node.loc;
  const { line, column } = location;

  return !(
        (start.line > line)
     || (start.line === line && start.column > column)
     || (end.line < line)
     || (end.line === line && end.column < column)
   );
}

function getPathClosestToLocation(source: SourceText, location: Location) {
  const ast = getAst(source);
  let pathClosestToLocation = null;

  traverse(ast, {
    enter(path) {
      if (nodeContainsLocation({ node: path.node, location })) {
        pathClosestToLocation = path;
      }
    }
  });

  return pathClosestToLocation;
}

function getVariablesInScope(source: SourceText, location: Location) {
  const path = getPathClosestToLocation(source, location);
  const bindings = get(path, "scope.bindings", {});

  return toPairs(bindings)
    .map(([name, binding]) => ({
      name,
      references: binding.referencePaths
    })
  );
}

module.exports = {
  parse,
  getSymbols,
  getPathClosestToLocation,
  getVariablesInScope,
  getExpression
};
