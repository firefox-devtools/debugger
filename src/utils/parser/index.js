// @flow

const { workerUtils: { WorkerDispatcher } } = require("devtools-utils");

const dispatcher = new WorkerDispatcher();

const getSymbols = dispatcher.task("getSymbols");
const getVariablesInScope = dispatcher.task("getVariablesInScope");
const resolveToken = dispatcher.task("resolveToken");

module.exports = {
  getSymbols,
  getVariablesInScope,
  resolveToken,
  startParserWorker: dispatcher.start.bind(dispatcher),
  stopParserWorker: dispatcher.stop.bind(dispatcher)
};
