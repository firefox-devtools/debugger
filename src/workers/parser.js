const babylon = require("babylon");
const traverse = require("babel-traverse").default;

import type { SourceText, Source } from "../types";

const ASTs = new Map();

function _parse(code) {
  return babylon.parse(code, {
    sourceType: "module",

    plugins: [
      "jsx",
      "flow"
    ]
  });
}

function parse(sourceText: SourceText, source: Source) {
  if (ASTs.has(source.id)) {
    return ASTs.get(source.id);
  }
  const ast = _parse(sourceText.text);
  ASTs.set(source.id, ast);
  return ast;
}

function getAst(source) {
  return ASTs.get(source.id);
}

function getFunctions(source) {
  const ast = getAst(source);

  traverse(ast, {
    enter(path) {}
  });

  return false;
}

module.exports = {
  parse,
  getFunctions
};
