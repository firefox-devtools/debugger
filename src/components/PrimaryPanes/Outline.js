// @flow

import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import classnames from "classnames";
import actions from "../../actions";
import { getSelectedSource, getSymbols } from "../../selectors";
import { isEnabled } from "devtools-config";
import "./Outline.css";
import previewFunction from "../shared/previewFunction";

import type {
  SymbolDeclarations,
  SymbolDeclaration
} from "../../utils/parser/getSymbols";
import type { SourceRecord } from "../../reducers/sources";

type State = any;
type Props = {
  isHidden: boolean,
  symbols: SymbolDeclarations,
  selectSource: (string, { line: number }) => void,
  selectedSource: ?SourceRecord
};

class Outline extends Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {};
  }

  selectItem(location) {
    const { selectedSource, selectSource } = this.props;
    if (!selectedSource) {
      return;
    }
    const selectedSourceId = selectedSource.get("id");
    const startLine = location.start.line;
    selectSource(selectedSourceId, { line: startLine });
  }

  renderFunction(func: SymbolDeclaration) {
    const { name, location } = func;

    return (
      <li
        key={`${name}:${location.start.line}:${location.start.column}`}
        className="outline-list__element"
        onClick={() => this.selectItem(location)}
      >
        {previewFunction({ name })}
      </li>
    );
  }

  renderFunctions() {
    const { symbols } = this.props;

    return symbols.functions
      .filter(func => func.name != "anonymous")
      .map(func => this.renderFunction(func));
  }

  render() {
    const { isHidden } = this.props;
    if (!isEnabled("outline")) {
      return null;
    }

    return (
      <div className={classnames("outline", { hidden: isHidden })}>
        <ul className="outline-list">
          {this.renderFunctions()}
        </ul>
      </div>
    );
  }
}

Outline.displayName = "Outline";

export default connect(
  state => {
    const selectedSource = getSelectedSource(state);
    return {
      symbols: getSymbols(state, selectedSource && selectedSource.toJS()),
      selectedSource
    };
  },
  dispatch => bindActionCreators(actions, dispatch)
)(Outline);
