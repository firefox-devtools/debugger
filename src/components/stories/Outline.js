/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import PropTypes from "prop-types";
import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";

import { Outline } from "../PrimaryPanes/Outline";
import { L10N } from "devtools-launchpad";

import "../App.css";
import "../PrimaryPanes/Outline.css";

require("devtools-modules/src/themes/dark-theme.css");

if (typeof window == "object") {
  window.L10N = L10N;
  window.L10N.setBundle(require("../../../assets/panel/debugger.properties"));
}

function makeFuncLocation(startLine) {
  return {
    start: {
      line: startLine
    }
  };
}

function makeSymbolDeclaration(name, line, klass = "") {
  return {
    id: `${name}:${line}`,
    name,
    location: makeFuncLocation(line),
    klass
  };
}

function OutlineFactory({ dir = "ltr", theme = "dark", ...props }) {
  const themeClass = `theme-${theme}`;
  document.body.parentNode.className = themeClass;
  return (
    <div
      className={`outline ${themeClass}`}
      dir={dir}
      style={{
        width: "300px",
        margin: "40px",
        border: "1px solid var(--theme-splitter-color)"
      }}
    >
      <Outline
        selectLocation={action("selectFrame")}
        isHidden={false}
        {...props}
      />
    </div>
  );
}

OutlineFactory.propTypes = {
  dir: PropTypes.string,
  theme: PropTypes.string
};

storiesOf("Outline", module)
  .add("empty", () => {
    const symbols = { functions: [], variables: [] };
    return <OutlineFactory symbols={symbols} />;
  })
  .add("simple", () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("foo", 2),
        makeSymbolDeclaration("render", 2)
      ],
      variables: []
    };
    return <OutlineFactory symbols={symbols} />;
  })
  .add("component", () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("foo", 2, "Fancy"),
        makeSymbolDeclaration("render", 2, "Fancy")
      ],
      variables: []
    };
    return <OutlineFactory symbols={symbols} />;
  })
  .add("complex component", () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("foo", 2, "Fancy"),
        makeSymbolDeclaration("render", 2, "Fancy"),
        makeSymbolDeclaration("other", 2),
        makeSymbolDeclaration("utilBar", 2)
      ],
      variables: []
    };
    return <OutlineFactory symbols={symbols} />;
  });
