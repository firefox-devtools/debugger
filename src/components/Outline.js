// @flow

import { DOM as dom, PropTypes, Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import actions from "../actions";
import { getSelectedSource, getSourceText } from "../selectors";
import { isEnabled } from "devtools-config";
import { getSymbols } from "../utils/parser";
import "./Outline.css";

import type { Record } from "../utils/makeRecord";
import type { SourceText } from "debugger-html";

class Outline extends Component {
  state: any;

  constructor(props) {
    super(props);
    this.state = {};
  }

  componentWillReceiveProps({ sourceText }) {
    if (sourceText) {
      this.setSymbolDeclarations(sourceText);
    }
  }

  async setSymbolDeclarations(sourceText: Record<SourceText>) {
    const symbolDeclarations = await getSymbols(sourceText.toJS());

    if (symbolDeclarations !== this.state.symbolDeclarations) {
      this.setState({
        symbolDeclarations
      });
    }
  }

  renderFunction(func) {
    return dom.li({ key: func.id }, func.value);
  }

  renderFunctions() {
    const { symbolDeclarations } = this.state;
    if (!symbolDeclarations) {
      return;
    }

    const { functions } = symbolDeclarations;

    return functions
      .filter(func => func.value != "anonymous")
      .map(this.renderFunction);
  }

  render() {
    if (!isEnabled("outline")) {
      return null;
    }

    return dom.div(
      { className: "outline" },
      dom.ul({}, this.renderFunctions())
    );
  }
}

Outline.propTypes = {
  selectedSource: PropTypes.object
};

Outline.displayName = "Outline";

export default connect(
  state => {
    const selectedSource = getSelectedSource(state);
    const sourceId = selectedSource ? selectedSource.get("id") : null;

    return {
      sourceText: getSourceText(state, sourceId)
    };
  },
  dispatch => bindActionCreators(actions, dispatch)
)(Outline);
