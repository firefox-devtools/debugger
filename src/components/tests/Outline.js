import React from "react";
import { shallow } from "enzyme";
const fromJS = require("../../utils/fromJS");
import Outline from "../../components/Outline";
import parser from "../../utils/parser";
import devtoolsConfig from "devtools-config";

const OutlineComponent = React.createFactory(Outline.WrappedComponent);

const sourcesToJs = fromJS({ text: "sources to js" });
const sourceText = {
  root: "some text here",
  toJS: function() {
    return sourcesToJs;
  }
};

describe("Outline", () => {
  var symbolDeclarations, symbolsPromise;

  beforeEach(() => {
    devtoolsConfig.isEnabled = jest.fn();
    devtoolsConfig.isEnabled.mockReturnValue(true);

    symbolDeclarations = {
      functions: [
        { id: "my_example_function1:21", value: "my_example_function1" },
        { id: "my_example_function2:22", value: "my_example_function2" }
      ]
    };

    parser.getSymbols = jest.fn();
    symbolsPromise = Promise.resolve(symbolDeclarations);
    parser.getSymbols.mockImplementation(args => {
      if (args == sourcesToJs) {
        return symbolsPromise;
      }
    });
  });

  it("should render a list of functions when properties change", async () => {
    const component = shallow(new OutlineComponent());

    component.setProps({ sourceText });

    await symbolsPromise;
    expect(component).toMatchSnapshot();
  });

  it("should render ignore anonimous functions", async () => {
    const component = shallow(new OutlineComponent());
    symbolDeclarations.functions[1] = {
      id: "anonymous:25",
      value: "anonymous"
    };

    component.setProps({ sourceText });

    await symbolsPromise;
    expect(component).toMatchSnapshot();
  });
});
