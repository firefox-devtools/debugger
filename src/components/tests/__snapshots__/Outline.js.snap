// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outline should render a list of functions when properties change 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="outline"
>
    <ul />
</div>,
  "nodes": Array [
    <div
      className="outline"
>
      <ul />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Outline
        sourceText={
                Object {
                        "root": "some text here",
                        "toJS": [Function],
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Outline {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "sourceText": Object {
            "root": "some text here",
            "toJS": [Function],
          },
        },
        "refs": Object {},
        "state": Object {
          "symbolDeclarations": Object {
            "functions": Array [
              Object {
                "id": "my_example_function1:21",
                "value": "my_example_function1",
              },
              Object {
                "id": "my_example_function2:22",
                "value": "my_example_function2",
              },
            ],
          },
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="outline"
>
          <ul>
                    <li>
                              my_example_function1
                    </li>
                    <li>
                              my_example_function2
                    </li>
          </ul>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="outline"
>
          <ul>
                    <li>
                              my_example_function1
                    </li>
                    <li>
                              my_example_function2
                    </li>
          </ul>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Outline
    sourceText={
        Object {
            "root": "some text here",
            "toJS": [Function],
          }
    }
/>,
}
`;

exports[`Outline should render ignore anonimous functions 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="outline"
>
    <ul />
</div>,
  "nodes": Array [
    <div
      className="outline"
>
      <ul />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Outline
        sourceText={
                Object {
                        "root": "some text here",
                        "toJS": [Function],
                      }
        }
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Outline {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "sourceText": Object {
            "root": "some text here",
            "toJS": [Function],
          },
        },
        "refs": Object {},
        "state": Object {
          "symbolDeclarations": Object {
            "functions": Array [
              Object {
                "id": "my_example_function1:21",
                "value": "my_example_function1",
              },
              Object {
                "id": "anonymous:25",
                "value": "anonymous",
              },
            ],
          },
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="outline"
>
          <ul>
                    <li>
                              my_example_function1
                    </li>
          </ul>
</div>,
        "_debugID": 4,
        "_renderedOutput": <div
          className="outline"
>
          <ul>
                    <li>
                              my_example_function1
                    </li>
          </ul>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Outline
    sourceText={
        Object {
            "root": "some text here",
            "toJS": [Function],
          }
    }
/>,
}
`;
